### PR TITLE
Improve audio recording performance

### PR DESCRIPTION
**Pull request by gsouquet**
Whilst continuing my quest for improved performance in the platform I stumbled upon some minor improvements in the audio recording part of the platform

This pull requests achieves two things
- Reduces the layout trashing by using `transform` over `height` to display the waveform, most browsers handle that at the compositing stage, https://csstriggers.com/
- Throttle the waveform rendering to ensure that it's never rendered at a faster rate than the screen refresh rate, using `requestAnimationFrame`

The waveform would receive on average 4 audio frame in a single tick, and would therefore process 3 more than it needs to

ping @turt2live as he is the person that probably knows the most about this bit of code

# Before 🐌 

Screen Shot 2021-06-22 at 12 23 21

# After 🐎 

Screen Shot 2021-06-22 at 12 21 50

### CODE REVIEW

File: src/components/views/rooms/VoiceRecordComposerTile.tsx [hunk 1]
```diff
@@ -178,8 +227,8 @@ export default class VoiceRecordComposerTile extends React.PureComponent<IProps,
 
         // only other UI is the recording-in-progress UI
         return <div className="mx_VoiceMessagePrimaryContainer mx_VoiceRecordComposerTile_recording">
-            <LiveRecordingClock recorder={this.state.recorder} />
-            <LiveRecordingWaveform recorder={this.state.recorder} />
+            <Clock seconds={this.state.seconds} />
+            <Waveform relHeights={this.state.relHeights} />
```
Comment (turt2live) [inline src/components/views/rooms/VoiceRecordComposerTile.tsx:231]: Can we restore the component structure here? It's fairly important for maintenance reasons (outside of this repo) to have the components be extremely independent.

File: res/css/views/voice_messages/_PlaybackContainer.scss [hunk 2]
```diff
@@ -33,9 +33,13 @@ limitations under the License.
     font-size: $font-14px;
     line-height: $font-24px;
 
+    contain: content;
+
     .mx_Waveform {
         .mx_Waveform_bar {
             background-color: $voice-record-waveform-incomplete-fg-color;
+            height: 100%;
+            transform: scaleY(max(0.05, var(--barHeight)));
```
Comment (turt2live) [inline res/css/views/voice_messages/_PlaybackContainer.scss:42]: A comment saying this variable comes from the JS side would be appreciated

**Review by turt2live — COMMENTED** [review]
overall the theory looks great - thanks for taking a look at this! Is part of the benefit from reducing the layout to less functional components, or can those safely be restored?

**Comment by gsouquet** [comment]
> Is part of the benefit from reducing the layout to less functional components, or can those safely be restored?

The amount of components does not make that much of a difference. Lifting the state to `VoiceRecordComposerTile` means that there's only one call to `setState` and that we do not have to duplicate that `requestAnimationFrame` throttling logic.

File: src/components/views/rooms/VoiceRecordComposerTile.tsx [hunk 3]
```diff
@@ -39,25 +46,67 @@ interface IProps {
...
+                // The incoming data is between zero and one, but typically even screaming into a
+                // microphone won't send you over 0.6, so we artificially adjust the gain for the
+                // waveform. This results in a slightly more cinematic/animated waveform for the
+                // user.
+                relHeights: bars.map(b => percentageOf(b, 0, 0.50)),
+                seconds: this.seconds,
+            });
+            this.scheduledAnimationFrame = false;
+        });
+    }
```
Comment (turt2live) [inline src/components/views/rooms/VoiceRecordComposerTile.tsx:108]: This all needs to move into the recording waveform too, as part of that component structure. I'm not seeing a reason why the clock needs to be counted as part of this, so should be fine to leave that one with the regular de-duped state updates?

Comment (turt2live) [reply src/components/views/rooms/VoiceRecordComposerTile.tsx:108]: To make this a bit more reusable in other components, we can leverage the `MarkedExecution` util we have, I think.

It needs a small update in order to work though: https://github.com/matrix-org/matrix-react-sdk/commit/824d92d22cecb1148a43b645ae61d83d57176250

Usage looks something like this: https://github.com/matrix-org/matrix-react-sdk/blob/ff02a41cfba89fd6454c8ec11a6e27a5b5f471d7/src/components/views/audio_messages/SeekBar.tsx#L50-L76

(these bits are part of an upcoming PR, but 824d92d22cecb1148a43b645ae61d83d57176250 should be cherrypickable)

File: src/components/views/voice_messages/LiveRecordingClock.tsx [hunk 4]
```diff
@@ -12,15 +12,62 @@ limitations under the License.
...
  */
 @replaceableComponent("views.voice_messages.LiveRecordingClock")
-export default class LiveRecordingClock extends React.PureComponent<IClockProps> {
+export default class LiveRecordingClock extends React.PureComponent<IProps, IState> {
+    private seconds = 0;
+    private rafId: number;
+
+    state = {
+        seconds: 0,
+    }
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingClock.tsx:41]: code style wants this in the constructor

File: src/components/views/voice_messages/LiveRecordingClock.tsx [hunk 5]
```diff
@@ -12,15 +12,62 @@ limitations under the License.
...
+        this.props.recorder.liveData.onUpdate((update: IRecordingUpdate) => {
+            this.seconds = update.timeSeconds;
+            this.scheduledUpdate.mark();
+        });
+    }
+
+    private scheduledUpdate = new MarkedExecution(
+        () => this.updateClock(),
+        () => this.onLiveDataUpdate(),
+    )
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingClock.tsx:53]: this should be above the function declarations, and surprising the linter isn't complaining about semicolons

File: src/components/views/voice_messages/LiveRecordingClock.tsx [hunk 6]
```diff
@@ -12,15 +12,62 @@ limitations under the License.
...
+
+    private onLiveDataUpdate() {
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingClock.tsx:55]: shouldn't need this function as it's not possible to overlap the animation frame request

File: src/components/views/voice_messages/LiveRecordingWaveform.tsx [hunk 7]
```diff
@@ -12,18 +12,66 @@ limitations under the License.
...
 @replaceableComponent("views.voice_messages.LiveRecordingWaveform")
-export default class LiveRecordingWaveform extends React.PureComponent<IWaveformProps> {
+export default class LiveRecordingWaveform extends React.PureComponent<IProps, IState> {
     public static defaultProps = {
         progress: 1,
     };
+
+    private waveform: number[] = [];
+    private rafId: number;
+
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingWaveform.tsx:42]: same comments from the clock apply to this file

File: src/utils/MarkedExecution.ts [hunk 8]
```diff
@@ -33,26 +33,33 @@ export class MarkedExecution {
     constructor(private fn: () => void, private onMarkCallback?: () => void) {
     }
 
+    /**
+     * Getter for the _marked property
+     */
+    public get marked() {
```
Comment (turt2live) [inline src/utils/MarkedExecution.ts:39]: doesn't look like this getter is used?

File: src/utils/MarkedExecution.ts [hunk 9]
```diff
@@ -22,34 +22,37 @@ limitations under the License.
  * The function starts unmarked.
  */
 export class MarkedExecution {
-    private marked = false;
+    private _marked = false;
```
Comment (turt2live) [inline src/utils/MarkedExecution.ts:25]: final nitpick: now that we're not using the property, it should be named without the underscore. 

```suggestion
    private marked = false;
```

File: src/utils/MarkedExecution.ts [hunk 10]
```diff
@@ -22,34 +22,37 @@ limitations under the License.
...
-    constructor(private fn: () => void) {
+    constructor(private fn: () => void, private onMarkCallback?: () => void) {
     }
 
     /**
      * Resets the mark without calling the function.
      */
     public reset() {
-        this.marked = false;
+        this._marked = false;
```
Comment (turt2live) [inline src/utils/MarkedExecution.ts:40]:
```suggestion
        this.marked = false;
```

File: src/utils/MarkedExecution.ts [hunk 11]
```diff
@@ -22,34 +22,37 @@ limitations under the License.
...
     }
 
     /**
      * Marks the function to be called upon trigger().
      */
     public mark() {
-        this.marked = true;
+        if (!this._marked) this.onMarkCallback?.();
+        this._marked = true;
```
Comment (turt2live) [inline src/utils/MarkedExecution.ts:48]:
```suggestion
        if (!this.marked) this.onMarkCallback?.();
        this.marked = true;
```

File: src/utils/MarkedExecution.ts [hunk 12]
```diff
@@ -22,34 +22,37 @@ limitations under the License.
...
     }
 
     /**
      * If marked, the function will be called, otherwise this does nothing.
      */
     public trigger() {
-        if (!this.marked) return;
+        if (!this._marked) return;
```
Comment (turt2live) [inline src/utils/MarkedExecution.ts:55]:
```suggestion
        if (!this.marked) return;
```

File: src/components/views/voice_messages/LiveRecordingWaveform.tsx [hunk 13]
```diff
@@ -15,46 +12,58 @@ limitations under the License.
...
-    public constructor(props) {
+    public static defaultProps = {
+        progress: 1,
+    };
+
+    private waveform: number[] = [];
+    private scheduledUpdate = new MarkedExecution(
+        () => this.updateWaveform(),
+        () => requestAnimationFrame(() => this.scheduledUpdate.trigger()),
+    )
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingWaveform.tsx:44]:
```suggestion
    );
```

File: src/components/views/voice_messages/LiveRecordingWaveform.tsx [hunk 14]
```diff
@@ -15,46 +12,58 @@ limitations under the License.
...
+
+    constructor(props) {
         super(props);
+        this.state = {
+            waveform: [],
+        }
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingWaveform.tsx:50]:
```suggestion
        };
```

File: src/components/views/voice_messages/LiveRecordingWaveform.tsx [hunk 15]
```diff
@@ -15,46 +12,58 @@ limitations under the License.
...
+} from "../../../voice/VoiceRecording";
 
 interface IProps {
-    recorder: VoiceRecording;
+    recorder?: VoiceRecording;
 }
 
 interface IState {
-    heights: number[];
+    waveform: number[]
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingWaveform.tsx:28]:
```suggestion
    waveform: number[];
```

File: src/components/views/voice_messages/LiveRecordingClock.tsx [hunk 16]
```diff
@@ -32,16 +33,31 @@ interface IState {
...
+        });
     }
 
-    private onRecordingUpdate = (update: IRecordingUpdate) => {
-        this.setState({seconds: update.timeSeconds});
-    };
+    private updateClock() {
+        this.setState({
+            seconds: this.seconds,
+        })
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingClock.tsx:59]:
```suggestion
        });
```

File: src/components/views/voice_messages/LiveRecordingClock.tsx [hunk 17]
```diff
@@ -32,16 +33,31 @@ interface IState {
...
+    private scheduledUpdate = new MarkedExecution(
+        () => this.updateClock(),
+        () => requestAnimationFrame(() => this.scheduledUpdate.trigger()),
+    )
+
+    constructor(props) {
         super(props);
+        this.state = {
+            seconds: 0,
+        }
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingClock.tsx:46]:
```suggestion
        };
```

File: src/components/views/voice_messages/LiveRecordingClock.tsx [hunk 18]
```diff
@@ -32,16 +33,31 @@ interface IState {
  */
 @replaceableComponent("views.voice_messages.LiveRecordingClock")
 export default class LiveRecordingClock extends React.PureComponent<IProps, IState> {
-    public constructor(props) {
+    private seconds = 0;
+    private scheduledUpdate = new MarkedExecution(
+        () => this.updateClock(),
+        () => requestAnimationFrame(() => this.scheduledUpdate.trigger()),
+    )
```
Comment (turt2live) [inline src/components/views/voice_messages/LiveRecordingClock.tsx:40]:
```suggestion
    );
```

**Review by turt2live — APPROVED** [review]
thanks for bearing with me on this. Going to apply the lints manually as I think you've probably had enough of me by now :p 

(also this is blocking other work I need to do)

**Pull request merged**